sklearn/utils/estimator_checks.py

```diff
@@ -1443,41 +1443,90 @@ def check_supervised_y_2d(name, estimator_orig):
+                               (classifier, ", ".join(map(str, y_exp)), ", "
+                               .join(map(str, y_pred))))
+
+    # training set performance
+    if name != "ComplementNB":
+        # This is a pathological data set for ComplementNB.
+        assert_array_equal(np.unique(y), np.unique(y_pred))
+    if np.any(classifier.classes_ != classes):
+        assert_array_equal(classes, classifier.classes_,
+                           err_msg="Unexpected classes_ attribute for %r: "
+                           "expected '%s', got '%s'" %
+                           (classifier, ", ".join(map(str, classes)), ", "
+                           .join(map(str, classifier.classes_))))
+
+
+def choose_check_classifiers_labels(name, y, y_names):
+    if name in ["LabelPropagation", "LabelSpreading"]:
+        # TODO some complication with -1 label
+        return y
+
+    return y_names
+
+
 def check_classifiers_classes(name, classifier_orig):
-    X, y = make_blobs(n_samples=30, random_state=0, cluster_std=0.1)
-    X, y = shuffle(X, y, random_state=7)
-    X = StandardScaler().fit_transform(X)
+    X_multiple, y_multiple = make_blobs(n_samples=30, random_state=0,
+                                        cluster_std=0.1)
+    X_multiple, y_multiple = shuffle(X_multiple, y_multiple, random_state=7)
+    X_multiple = StandardScaler().fit_transform(X_multiple)
     # We need to make sure that we have non negative data, for things
     # like NMF
-    X -= X.min() - .1
-    X = pairwise_estimator_convert_X(X, classifier_orig)
-    y_names = np.array(["one", "two", "three"])[y]
+    X_multiple -= X_multiple.min() - .1
 
-    for y_names in [y_names, y_names.astype('O')]:
-        if name in ["LabelPropagation", "LabelSpreading"]:
-            # TODO some complication with -1 label
-            y_ = y
-        else:
-            y_ = y_names
+    X_binary = X_multiple[y_multiple != 2]
+    y_binary = y_multiple[y_multiple != 2]
 
-        classes = np.unique(y_)
-        classifier = clone(classifier_orig)
-        if name == 'BernoulliNB':
-            X = X > X.mean()
-        set_random_state(classifier)
-        # fit
-        classifier.fit(X, y_)
+    X_multiple = pairwise_estimator_convert_X(X_multiple, classifier_orig)
+    X_binary = pairwise_estimator_convert_X(X_binary, classifier_orig)
 
-        y_pred = classifier.predict(X)
-        # training set performance
-        if name != "ComplementNB":
-            # This is a pathological data set for ComplementNB.
-            assert_array_equal(np.unique(y_), np.unique(y_pred))
-        if np.any(classifier.classes_ != classes):
-            print("Unexpected classes_ attribute for %r: "
-                  "expected %s, got %s" %
-                  (classifier, classes, classifier.classes_))
+    labels_multiple = ["one", "two", "three"]
+    labels_binary = ["one", "two"]
+
+    y_names_multiple = np.array(labels_multiple)[y_multiple]
```

Comment:
this could be written more explicitly and clearly as `np.take(labels_multiple, y_multiple)`